pkg/tinyvcdiff/contrib/tinyvcdiff_mtd/mtd.c

```diff
@@ -0,0 +1,179 @@
+/*
+ * Copyright (C) 2022 Juergen Fitschen
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ */
+
+#include "vcdiff_mtd.h"
+#include "assert.h"
+#include <string.h>
+
+#define ENABLE_DEBUG 0
+#include "debug.h"
+
+static int _erase (void *dev, size_t offset, size_t len)
+{
+    int rc;
+    vcdiff_mtd_t *mtd = dev;
+    size_t addr = offset + len;
+    size_t page = addr / mtd->dev->page_size;
+    size_t sector = page / mtd->dev->pages_per_sector;
+    size_t cnt = sector - mtd->next_erase_sector + 1;
+
+    /* early exit if all sectors are already in the erased state */
+    if (cnt == 0) {
+        return 0;
+    }
+
+    /* erase sectors */
+    rc = mtd_erase_sector(mtd->dev, mtd->next_erase_sector, cnt);
+    if (rc != 0) {
+        return rc;
+    }
+
+    mtd->next_erase_sector += cnt;
+
+    return 0;
+}
+
+static int _read (void *dev, uint8_t *dest, size_t offset, size_t len)
+{
+    vcdiff_mtd_t *mtd = dev;
+
+    /* mtd_read_page will take care of calculation the right page */
+    int rc = mtd_read_page(mtd->dev, dest, 0, offset, len);
+
+    return rc;
+}
+
+static int _align_write (vcdiff_mtd_t *mtd, uint8_t **src, size_t *len)
+{
+    size_t alignment_offset;
+    size_t copy_len;
+
+    alignment_offset = mtd->offset % CONFIG_TINYVCDIFF_MTD_WRITE_SIZE;
```

Comment:
`vcdiff_mtd_t` requires `write_size` to be known at compile time to adjust buffer sizes. But at least an `assert()` would be nice to have.

Reply:
Could we use `mtd->work_area` instead?

Reply:
This would add the dependency `mtd_write_page`. Currently, I'm not using any of the write page features. Adding this module would add unnecessary code just to have the (`mtd_write_page`-internal) buffer at head. It feels a little bit hacky to me.

I'd say having `assert()`s in place that make sure the buffer is aligned and large enough would be a good thing to have in place.

The default of 4 bytes seems to be a size that fits all `mtd` devices available in RIOT (IIRC) and isn't wasteful in my opinion.

Reply:
For `mtd_flashpage` it can be larger than 4, same for SD cards (but those are usually used with a filesystem)